### ISSUE TITLE
Introduce message component(s)

### ISSUE THREAD

**Issue by PascalPrecht** [status-im/StatusQ#499]
This issue serves a ground to discuss the new API(s) for message components provided by StatusQ. 
The discussion has originally started in https://github.com/status-im/status-desktop/issues/4272.

Comments in the original issue have been echo'd here.

- [x] create skeleton with profile image and dummy model
- [x] https://github.com/status-im/StatusQ/issues/501
- [x] https://github.com/status-im/StatusQ/issues/502
- [x] https://github.com/status-im/StatusQ/issues/505
- [x] https://github.com/status-im/StatusQ/issues/506
- [x] https://github.com/status-im/StatusQ/issues/504
- [x] https://github.com/status-im/StatusQ/issues/507
- [x] https://github.com/status-im/StatusQ/issues/508
- [x] https://github.com/status-im/StatusQ/issues/511
- [x] https://github.com/status-im/StatusQ/issues/512
- [ ] invitation bubble
- [ ] links
- [ ] transaction
- [ ] time info ?
- [ ] Audio Message new design with audio graph

**Comment by PascalPrecht** [comment]
[Comment](https://github.com/status-im/status-desktop/issues/4272#issuecomment-992500744) by @Khushboo-dev-cpp 


This issue needs to be analysed and be divided into several subtasks -

API should be something like below

1. Sender Name
2. has mention
3. Time info
4. Profile Picture info
5. Pinned message
6. Message Content (Create separate Components for each of the below in StatusQ
  a. Text message
  b. Image message (open in full screen should be self contained and also the various actions of right click of button?)
  c. Sticker
  d. Gif
  e. Audio
7. Edit Mode
8. is a reply and the reply component itself.
9. Quick Actions button panel

**Comment by PascalPrecht** [comment]
[Comment](https://github.com/status-im/status-desktop/issues/4272#issuecomment-992754961) by @Khushboo-dev-cpp 

Noting down here is an initial version of what the API for the Message component should look like

```qml
StatusMessage {
    messageFromUser: true
    senderName: "Sender of Message"
    senderProfilePicture: image||identicon
    contentType: StatusMessage.Type.Text ||
                 StatusMessage.Type.Image ||
                 StatusMessage.Type.Sticker ||
                 StatusMessage.Type.Audio
    messageText: "This is a message"
    imageMsgSource: "img"
    stickerData: sticker
    timeInfo: ""
    isMessagePinned: true
    pinnedBy: someone
    isAReply: true
    // replyDetails: Can convert this into a sort of array of vars but maybe difficult to read
    senderNameBeingReplied: "senderName"
    senderProfilePictureBeingReplied: "source"
    contentTypeBeingReplied: StatusMessage.Type
    msgTextBeingReplied: text
    imageMsgSourceBeingReplied: "image"
    stickerBeingReplied: sticker
    editMode: false
    // Edit mode can be handled completely internally with a signal emitted to set new text
    onEditCompleted: {
    // inform BE to update sent Message
    }
    quickActions: {StatusFlatRoundButton{onClicked: {/*Do Action*/}}, StatusFlatRoundButton{},StatusFlatRoundButton{}}
    // Should EMoji reactions Panel or EMoji be in StatusQ? what if this expands to stickers in future?
    footerComponent: EmojiReactionsPanel
     // We have constraints that we cant show message + Image in same message? Will there be support in the future? should we create a component that supports that already?
    onRightClick: {/* show a different context menu*/}
}

// Create components for the content Items, 
```
AudioMessage
Image with its fullscreen overlay self contained
ReplyPanel
this isn't a 100% complete, I think this will have to be updated as we move forward

**Comment by PascalPrecht** [comment]
[Comment](https://github.com/status-im/status-desktop/issues/4272#issuecomment-992985105) by @alexandraB99 

This component is generally used as a delegate thus getting all info from a model. In case of ChatMessagesView and PinnedMessagesPopup is model.property and in the case of ActivityCenterMessageComponentViewis model.message.property. Exception is the case that is used in AppearanceView as preview sample where the properties are set with a default value. That said, we could probably directly do all assignments inside the component instead of exposing all these properties as an API and have it detecting where is in ActivityCenterMessageComponentView to get from model.message otherwise just model. In the AppearanceView case we could either think about creating a minimal MessageView component that only demonstrates a preview or use the same one in a "preview" mode. What we should really think about is expose somehow several of its properties that are needed by other components, it could be some function that returns the specific property from the specific selected message in the view or something like this.

**Comment by PascalPrecht** [comment]
Thanks @Khushboo-dev-cpp for kicking this off! Happy to see some more concrete ideas of what this could look like and I think the direction is good.

Below are a few comments/thoughts/questions:

1. Can you clarify what `messageFromUser` is for? It **seems** like we have messages that aren't generated/sent by users but something else? Maybe you can add a comment to that property that clarifies this...

2. How do you feel about introducing a `StatusMessage.ContentType` enum for `contentType`, instead of using a generalized `Type`? Reason I'm suggesting this, is that we also have these "transaction bubbles" which are also messages. In such case we might need a `Type.Transaction` or something like that. Although, we could also have an entirely separate component for that since it's very different from a "standard" message. Curious what you think

3. What's the shape of `stickerData`?

4. Is `timeInfo` going to be the timestamp/date of the message? Maybe we can make it an actual `timestamp`

5. I like the idea of attaching the `replyDetails` to the message. How do you feel about making that whole thing an object that we can dot into? Then we don't need to increase the API service with all those `..beingRepliedTo` properties. So for example:

```
replyDetails.senderName: "..."
replyDetails.senderProfilePicture: "..."
etc
```

Then, we can probably also get rid off the `isReply` property because we can use heuristics for that (if we have `replyDetails` we know we're dealing with a reply.

6. I **assume** `quickActions` is the popup menu with buttons for editing, reacting etc? In that case, I think we can use a similar to pattern that we've used in many other StatusQ components and pass a `StatusPopupMenu` (or a variation of that) to the component. That will also give consumers of the component full control over what options/buttons should be available. Then, based on the existence of the menu we know whether we need to instantiate one in the first place.

7. Okay, realising just now what I've described in 6) is what you were aiming for with `footerComponent` ? How exactly is that a footer and what is `quickActions` for then?

8. Wonder if, instead of having `onRightClick`, we just have a plain old `onClick` handler that propagates the `mouseEvent` with information attached to it (like whether it was a right click or not).

Let me know what you think!

Once we have settled on some of those, we can determine what other things are missing (for example ENS name handling etc).


> What we should really think about is expose somehow several of its properties that are needed by other components, it could be some function that returns the specific property from the specific selected message in the view or something like this.

@alexandraB99 would you mind elaborating on this? I don't follow what you mean by "exposing several of its properties". Isn't that exactly what Khushboo's proposal is doing?

**Comment by Khushboo-dev-cpp** [comment]
> his component is generally used as a delegate thus getting all info from a model. In case of ChatMessagesView and PinnedMessagesPopup is model.property and in the case of ActivityCenterMessageComponentViewis model.message.property. Exception is the case that is used in AppearanceView as preview sample where the properties are set with a default value. That said, we could probably directly do all assignments inside the component instead of exposing all these properties as an API and have it detecting where is in ActivityCenterMessageComponentView to get from model.message otherwise just model. In the AppearanceView case we could either think about creating a minimal MessageView component that only demonstrates a preview or use the same one in a "preview" mode. What we should really think about is expose somehow several of its properties that are needed by other components, it could be some function that returns the specific property from the specific selected message in the view or something like this.

Hey @alexandraB99 I thought about passing the model directly to. The only issue with designing a reusable widget this way would be and we have recently experienced this is that if the BE changes the widget needs an update too. and this doesn't seem ideal. Our widget should be easy to use irrespective of the BE API. Let discuss this further in case there is a better way to solve this.

**Comment by PascalPrecht** [comment]
> Shape? Type would be string.

That's good to know! So if it's a string, and we have the `contentType`, if every content type can be specified as string, we can even explore getting rid of `imageMsgSource`, `stickerData`, `audioSource`(?), etc and have a one `payload` property instead? We could even put `messageText` in there as well then. Although, for usability/convenience, I'd keep that one around.

Anyways, just an idea.

>  I have seen this being used in our code and found it difficult to use and assign

Oh I see.. Not trying to convince you here.. QML built-in APIs do the same thing in some places (e.g. `icon.name`, `icon.width`, `icon.source`) etc.  If it's hard to use then it's probably not worth it. Should be clear **why** it's hard to use. Another way to assign those values would be via:

```
someProp: QtObject {
  foo: ...,
  bar: ...
}
```

Would that make it simpler?

> So quick actions is the buttons bar that appears on hover of the message

Ah right. Yea in that case `quickActions` is probably doing the trick. Just need to decide how we go about loading/unloading those actions. For example, if I want to render a message component without quick actions (e.g. message preview component in the appearance view), what would that look like? Is that an empty array of buttons? Is it an array of buttons but all are disabled or invisible?

> I added this footer for the emoji reactions that can be added to the Message,called that footer so that its generic and can be used to add something else in the future? There will only. be leaders available in the StatusMessage to load these components, unknown to the StatusMessage what these will contain.

I like the idea.. I think we should probably keep a convenience API then, which will effectively be a shorthand for `footerComponent`. So for example, at the moment there's only reactions. We might have a dedicated `reactions` property which will hydrate the `footerComponent`, however there's also `footerComponent` which can be used instead in case one needs more control.

Wdyt?

**Comment by Khushboo-dev-cpp** [comment]
> > Shape? Type would be string.
> 
> That's good to know! So if it's a string, and we have the `contentType`, if every content type can be specified as string, we can even explore getting rid of `imageMsgSource`, `stickerData`, `audioSource`(?), etc and have a one `payload` property instead? We could even put `messageText` in there as well then. Although, for usability/convenience, I'd keep that one around.
> 
> Anyways, just an idea.
> 

yes one type something like messageContent. We can try this out. 

> > I have seen this being used in our code and found it difficult to use and assign
> 
> Oh I see.. Not trying to convince you here.. QML built-in APIs do the same thing in some places (e.g. `icon.name`, `icon.width`, `icon.source`) etc. If it's hard to use then it's probably not worth it. Should be clear **why** it's hard to use. Another way to assign those values would be via:
> 
> ```
> someProp: QtObject {
>   foo: ...,
>   bar: ...
> }
> ```
> 
> Would that make it simpler?
> 

Do you mean the way we have exposed IconSettings and ImageSettings? Thats actually very cool. What I meant was creating an Object like var replyDetails: {name: :"", message: "", image: ""}
The way you described it with QtObject is cool too.

> > So quick actions is the buttons bar that appears on hover of the message
> 
> Ah right. Yea in that case `quickActions` is probably doing the trick. Just need to decide how we go about loading/unloading those actions. For example, if I want to render a message component without quick actions (e.g. message preview component in the appearance view), what would that look like? Is that an empty array of buttons? Is it an array of buttons but all are disabled or invisible?
> 
Maybe better to have them invisible?

> > I added this footer for the emoji reactions that can be added to the Message,called that footer so that its generic and can be used to add something else in the future? There will only. be leaders available in the StatusMessage to load these components, unknown to the StatusMessage what these will contain.
> 
> I like the idea.. I think we should probably keep a convenience API then, which will effectively be a shorthand for `footerComponent`. So for example, at the moment there's only reactions. We might have a dedicated `reactions` property which will hydrate the `footerComponent`, however there's also `footerComponent` which can be used instead in case one needs more control.
> 
> Wdyt?

sure..makes sense! Same for quick Actions I guess then?

**Comment by PascalPrecht** [comment]
> The way you described it with QtObject is cool too.
So you mentioned you found that style of property assignment difficult to use (namely that you do `foo.bar: baz` instead of `fooBar: baz`). I was trying to figure out what the difficulty was and whether using `foo: QtObject { bar: baz }` would be "easier".

> Maybe better to have them invisible?

So say I give it a list of `quickActions`: `StatusFlatButton { ... }, StatusFlatButton { ... }`. In this case it's easy to just hydrate the component with these actions/buttons.  In other cases I might not want to have them at all (e.g.  message preview mentioned by @alexandraB99 ), so then I guess `quickActions` is just gonna be empty.

What if there's **some** of them enabled, others disabled? We just need to ensure we detect these changes to reflect their (dis)appearance in the UI.

> Same for quick Actions I guess then?

We could, although I don't see a need for it at the moment (actually very similar to `footerComponent`, but it sounded like you know about some possible other footer components haha)

**Comment by Khushboo-dev-cpp** [comment]
> > The way you described it with QtObject is cool too.
> > So you mentioned you found that style of property assignment difficult to use (namely that you do `foo.bar: baz` instead of `fooBar: baz`). I was trying to figure out what the difficulty was and whether using `foo: QtObject { bar: baz }` would be "easier".
> 
> > Maybe better to have them invisible?
> 
> So say I give it a list of `quickActions`: `StatusFlatButton { ... }, StatusFlatButton { ... }`. In this case it's easy to just hydrate the component with these actions/buttons. In other cases I might not want to have them at all (e.g. message preview mentioned by @alexandraB99 ), so then I guess `quickActions` is just gonna be empty.
> 
> What if there's **some** of them enabled, others disabled? We just need to ensure we detect these changes to reflect their (dis)appearance in the UI.
> 
I get it now. Yes then empty when they are not needed at all and invisible when they are available conditionally.

> > Same for quick Actions I guess then?
> 
> We could, although I don't see a need for it at the moment (actually very similar to `footerComponent`, but it sounded like you know about some possible other footer components haha)

haha no I was just think from a future perspective.

**Comment by Khushboo-dev-cpp** [comment]
Updating the API for Status Message after completing implementation of the statusMessageHeader 
which represents the 
![image](https://user-images.githubusercontent.com/60327365/146951233-ad5896a4-b8aa-4c8e-ae7e-e4da8e6ba50c.png)

StatusMessage {
   // is the current message coming from the user itself
    messageFromUser: true
// Part of StatusMessageHeader
                 messageHeader.displayName: model.userName
                 messageHeader.secondaryName: model.ensName
                 messageHeader.tertiaryDetail: model.chatKey
                 messageHeader.icon1.name: ""
                 messageHeader.icon2.name: ""
                 timeStamp: "10:00 am"
// Profile Image
                 senderProfilePicture.source: model.profileImage
                 senderProfilePicture.isIdenticon: model.isIdenticon
    contentType: StatusMessage.ContentType.Text ||
                 StatusMessage.ContentType.Image ||
                 StatusMessage.ContentType.Sticker ||
                 StatusMessage.ContentType.Audio ||
                 StatusMessage.ContentType.Transaction
    messageText: "This is a message"
    imageMsgSource: "img"
    stickerData: "sticker"
    timeStamp: ""
    isPinned: true
    pinnedBy: someone
    isAReply: true
    // replyDetails: Can convert this into a sort of array of vars but maybe difficult to read
    senderNameBeingReplied: "senderName"
    senderProfilePictureBeingReplied: "source"
    contentTypeBeingReplied: StatusMessage.Type
    msgTextBeingReplied: text
    imageMsgSourceBeingReplied: "image"
    stickerBeingReplied: sticker
    editMode: false
    // Edit mode can be handled completely internally with a signal emitted to set new text
    onEditCompleted: {
    // inform BE to update sent Message
    }
    quickActions: {StatusFlatRoundButton{onClicked: {/*Do Action*/}}, StatusFlatRoundButton{},StatusFlatRoundButton{}}
    // Should EMoji reactions Panel or EMoji be in StatusQ? what if this expands to stickers in future?
    footerComponent: EmojiReactionsPanel
     // We have constraints that we cant show message + Image in same message? Will there be support in the future? should we create a component that supports that already?
    onClicked: {/* show a different context menu*/}
}

**Comment by Khushboo-dev-cpp** [comment]
Things still pending to be done are listed below -
Created only a loader in the StatusMessage to load the -
1. invitation, 
2. links and 
3. transaction message content Type.
Somehow feels like something the user of the component can define for themselves. In case after a discussion with the team it seems necessary to add it to the StatusMessage Component OR move the Invitation, Links and Transaction containers to StatusQ, we would need to create sub tasks for those.
4. Also pending is the audio graph in an Audio message  in the new design. (the logic for the graphics bar needs to be discussed)
5. Need to discuss further about 
            a. Channel Identifier
            b. FetchMoreMessagesButtonComponent
            c.  PrivateGroupHeaderComponent
            d. Gap Component
The usage of this in desktop may bring about some changes in this StatusQ component as it has been designed with it being a purely design based UI component so far. 

Will be shorting raising a review for the rest of it that has been completed

**Comment by PascalPrecht** [comment]
Closing this as 80% of this has landed as #500
